### ISSUE TITLE
Potential issue with unawaited UpdateClientAssertion event

### ISSUE THREAD

**Issue by ovska** [IdentityModel/IdentityModel.AspNetCore.OAuth2Introspection#160]
https://github.com/IdentityModel/IdentityModel.AspNetCore.OAuth2Introspection/blob/main/src/OAuth2IntrospectionHandler.cs#L211

The `OnUpdateClientAssertion` event is invoked without an `await`, probably because it is inside a `lock`. This can potentially cause hard to diagnose bugs if the event is assigned to something that requires awaiting. For example the current HttpContext is one of the properties, and using it after the request is ended can throw object disposed exceptions.

My suggestion is to replace the lock with a `SemaphoreSlim` which works fine in async context as well, and the lifetime of `OAuth2IntrospectionOptions` shouldn't be a problem as [SemaphoreSlim wouldn't need to be disposed in this case](https://stackoverflow.com/a/39195920).

Tangentially, `ConfigureAwait(false)` is missing from some all awaits on events. Is this intentional or an oversight?

**Comment by leastprivilege** [comment]
My understanding is that ASP.NET has no synchronization context - so `ConfigureAwait(false)` is not required.

Would you agree?

**Comment by ovska** [comment]
Correct, was just wondering why it was used in some awaits and not all :)

**Comment by leastprivilege** [comment]
I guess because it is ultimately an old code base..

**Comment by leastprivilege** [comment]
@brockallen could you have a look when time permits. 

@ovska do you want to propose a PR?

**Comment by ovska** [comment]
Here's more or less the minimal fix I could come up with: #164
